content/blog/2021/hidden_powers_of_bitcoin.md

```diff
@@ -0,0 +1,508 @@
+
+Here are some fragments which can be combined to create powerful expressions.
+
+1. `pk(key)` - Specifies a given public key
+2. `thresh(k, expr_1, expr_2, ..., expr_n)` - Specifies k of n multisig using expressions.
+3. `older(T)` - Timelock for T blocks
+4. `and(expr_1, expr_2)` - Both expressions should evaluate to true. 
+5. `or(expr_1, expr_2)` - Any one of the expressions should evaluate to true.
+6. `aor(expr_1, expr_2)` - Similar to `or` but `expr_1` has a more probability to evaluate to true.
+
+Bitcoin Script allows us to use another alternate stack. The combinator functions use this second stack to evaluate expressions of `thresh`, `and`, `aor` and `or`.
+The complete Miniscript Reference can be found [here](http://bitcoin.sipa.be/miniscript/#:~:text=Miniscript%20reference).
+
+##### Example Policies
+
+Here are the Miniscript Policies for the examples we looked at earlier. 
+Note `A`, `B`, `C` are placeholders for keys *(or output descriptors)* involved in the tx.
+
+1. Pay A (pay-to-public-key)
+```
+pk(A)
+```
+
+2. Escrow Account between A, B and third-party C.
+```
+thresh(2,pk(A),pk(B),pk(C))
+```
+
+3. Vault for A time-locked for T blocks with B as the master key.
+```
+aor(and(pk(A),time(T)),pk(B))
+```
+
+4. HTLC payment to B, which, if unspent for T blocks, returns to A.
+```
+aor(and(pk(A),time(T)),and(pk(B),hash(H))))
+```
+
+The Miniscript Policy Compiler is written in Rust and is present in [this repository](https://github.com/rust-bitcoin/rust-miniscript). 
+In this blog, we will later use the same using [bitcoindevkit/bdk](https://github.com/bitcoindevkit/bdk), a lightweight descriptor-based wallet library
+with a [cli](https://github.com/bitcoindevkit/bdk-cli). 
+
+### Descriptors
+
+Descriptors are a compact and efficient way to "describe" how scripts *(and addresses)* of a wallet should be generated.
+They make it easier to deal with Multisig or complicated key setups.
+Descriptors are super portable and can be easily used by any wallet to determine the list of all addresses that can be generated from the same.
+This feature creates a common stage for all Bitcoin apps and software.
+It's an abstract way to replace keys.
+
+According to Bitcoin Core, Output Descriptors are "a simple language which can be used to describe collections of output scripts".
+They bring in themselves, derivation paths, master xpub/xprv fingerprints and paths to generate addresses from.
+Let's understand this with an example of an Output Descriptor:
+
+```
+
+Descriptor: pkh([d34db33f/44'/0'/0']xpub6ERaJH[...]LJRcEL/1/*)#ml40v0wf
+            <1> <--------2---------><----------3---------><4> <---5--->
+
+Sections:
+1 - function type (here, P2PK)
+2 - derivation path of this key
+3 - xpub at m/44'/0'/0
+4 - path to deriving keys/addresses at
+5 - checksum for the descriptor
+```
+
+This was an example of a simple descriptor using one function but they can also be used in conjugation with multiple functions like `combo`, `multi` etc.
```

Comment:
`combo` and `multi` are two different things. `combo` is a descriptor which describes scriptpubkey of any type ex: P2PKH, P2SH etc.

`multi` is basically a miniscript which describes a m-of-n multisig script. When used as descriptor it will make the scriptpubkey as the raw multisig script. But its better to think it as a miniscript than a descriptor, because such raw multisigs are never used in practice.

In wild you will see stuffs like `wsh(multi(2, PKA, PKB, PKC))`. The `wsh` part is the descriptor describing its a P2WSH output, and `multi` part is the miniscript describing the actual spending condition.

Ya it gets messy with terminology in bitcoin scripts. But if we aren't careful, this can cause more confusion to readers.

Reply:
True, I added this section in the blog because it makes the concept clearer.